src/bwt_util.rs

```diff
@@ -1,3 +1,111 @@
+use bitvec::prelude::*;
+use itertools::Itertools;
+use std::clone::Clone;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::hash::Hash;
+
+/// implementation of the mergeIter function as described in Holt & McMillan 2014
+/// Takes the current state of the interleave bitvector, the two burrows wheeler
+/// transform slices, and a mapping from character to number of lexigraphically
+/// smaller characters in the BWTs provided and does a round of interleave
+/// resolution
+pub fn pairwise_merge_iter<T: Ord + Hash + Clone + Debug>(
+    current_interleave: &BitVec<u64, Msb0>,
+    bwt0: &[T],
+    bwt1: &[T],
+    offsets: &HashMap<T, usize>,
+) -> BitVec<u64, Msb0> {
+    // initial conditions
+    let mut next_interleave = bitvec![u64, Msb0; 0; current_interleave.len()];
+    let mut current_pos0 = 0usize;
+    let mut current_pos1 = 0usize;
+
+    let mut temp_index = offsets.clone();
+
+    // iterate through each bit
+    for b in current_interleave.iter() {
+        let c = if *b {
+            let v = bwt0[current_pos0].clone();
+            current_pos0 += 1;
+            v
+        } else {
+            let v = bwt1[current_pos1].clone();
+            current_pos1 += 1;
+            v
+        };
+        // copy b into the next position for c
+        let next_position = temp_index
+            .get_mut(&c)
+            .expect("Unknown character encountered in merging BWTs");
+        next_interleave.set(*next_position, *b);
+        // update the tempIndex to match the FM-index
+        *next_position += 1;
+    }
+
+    next_interleave
+}
+
+/// Determines the offset hashmap for a provided slice of BWTs.
+/// The offset hashmap is the mapping from a character to the number of
+/// lexigraphically lesser characters present in the combined string provided.
+fn generate_offset_hashmap<T: Ord + Hash + Clone + Debug>(bwts: &[&[T]]) -> HashMap<T, usize> {
+    let mut num_occurrences: HashMap<T, usize> = HashMap::new();
+    for &bwt in bwts.iter() {
+        for c in bwt {
+            if let Some(i) = num_occurrences.get_mut(c) {
```

Comment:
Minor, but I think this may be more rust-like?
```
let entry = num_occurrences.entry(c).or_insert(0);
*entry += 1;
```

Reply:
I would guess that either the HashMap key needs to be &T or T needs to also implement the Copy trait.  They use this exact thing in the `entry` example, so seems like a prime place: https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html#method.entry

With that said, it's not that important, just seemed like an easy code reduction.